UI: Design a 404 page for default redirects

Aaishpra opened the pull request:
### Description
Solves the Following.

   

-  Design a 404 page

   

-  Change route URL to Redirect to /404 on wrong URLs

Fixes #33

### Type of Change:
<!--**Delete irrelevant options.**-->

- Code

- User Interface


**Code/Quality Assurance Only**

- New feature (non-breaking change which adds functionality pre-approved by mentors)



### How Has This Been Tested?
<!--Describe the tests you ran to verify your changes. Provide instructions or GIFs so we can reproduce. List any relevant details for your test.-->
ScreenShot for Reference
![image](https://user-images.githubusercontent.com/66299533/91637077-3c05f600-ea23-11ea-9c95-f227be102d04.png)


### Checklist:
<!--**Delete irrelevant options.**-->

- [x]  My PR follows the style guidelines of this project

- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] Any dependent changes have been merged
- [x] I have written Kotlin Docs whenever is applicable


**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [x] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules


**Additional Context**
 I initially wanted to add a button to go back to the home page but since it doesn't Exist hence skipped that part. More suggestions and Additions are welcome for any other Feature within the page

dependabot[bot] committed:
chore: bump elliptic from 6.5.2 to 6.5.3 (#46)

Created File ErrorPage.js for Error page

Create ErrorPage.css

Update Routes.js

Changed Route to AuthRoute

Removed Navbar and Extra Classes from ErrorPage.js

Updated ErrorPage.css

Changed Defualt function to Class component

Update ErrorPage.css

dependabot[bot] committed:
chore: bump elliptic from 6.5.2 to 6.5.3 (#46)

Created File ErrorPage.js for Error page

Create ErrorPage.css

Update Routes.js

Changed Route to AuthRoute

Removed Navbar and Extra Classes from ErrorPage.js

Updated ErrorPage.css

Changed Defualt function to Class component

Update ErrorPage.css

bismitaguha commented:
You should have used AuthRoutes IMO!

Aaishpra commented:
Done @bismitaguha please review

bismitaguha commented on src/components/ErrorPage.js:
```diff
@@ -0,0 +1,24 @@
+
+function ErrorPage() {
+    return (
+        <>    <Navbar /> 
```
AuthRoute already contains the Navbar, so this wasn't needed. Reusing components is always preferable. That's why I suggested AuthRoute.

bismitaguha commented on src/components/ErrorPage.js:
```diff
@@ -0,0 +1,24 @@
+    )
+}
+
+export default ErrorPage
```
Throughout the application class method has been used.

bismitaguha replied:
@Aaishpra Make these changes

Aaishpra replied:
Already done @bismitaguha

Aaishpra replied:
@bismitaguha is this correct now

bismitaguha replied:
Yes

Aaishpra commented:
Changed the files @bismitaguha

Aaishpra commented:
@bismitaguha please review the changes

Aaishpra commented:
@abha224 can u review the pull request as @bismitaguha seems not to be available

bismitaguha commented:
I had requested changes @Aaishpra

Aaishpra commented:
@bismitaguha can u please check my latest commits i guess i changed the file as u requested.

bismitaguha commented:
@Aaishpra I didn't ask you to remove classes. The code already written is in exporting a class component not by exporting default function methods
Just go through the `src/` directory once.

sidvenu commented:
@Aaishpra is the 404 text color the same as the color of Open Source Programs header? If not, I suggest switching to that color.

Aaishpra commented:
@bismitaguha does it needs to be done by using class ErrorBoundary 
@sidvenu the color is the same

Aaishpra commented:
@sidvenu the color is the same maybe  it looks a bit different because the text 404 is much bolder than Open source programs

Aaishpra commented:
@bismitaguh is the latest commit ok now?

Aaishpra commented:
@bismitaguha please review the changes

Aaishpra commented:
No Problem @bismitaguha got your point.

Aaishpra commented:
@sidvenu can you review the PR

sidvenu reviewed (dismissed):
LGTM!

sidvenu commented:
@anitab-org/qa-team @anitab-org/open-source-progs-maintainers  @anitab-org/open-source-progs-developers Can anyone test this PR out?

Aaishpra commented:
Hey,@bismitaguha and @sidvenu can you suggest someone that can test this PR out so that i can personally tag them here, or you people can help with it. Actually a lot of time has already passed since the last review, and no one has come to see this here.

sidvenu commented:
@rpattath @bismitaguha if possible, could you test this PR?

devkapilbansal reviewed (changes requested):
@Aaishpra please squash your commits into one. It's always better to have one commit per change to keep the history clean

devkapilbansal commented on src/styles/ErrorPage.css:
```diff
@@ -0,0 +1,44 @@
+    color: #0d0d0d;
+    
+}
+
+
```
Why two blank lines are here? We only need one

Aaishpra commented:
@devkapilbansal can you tell me what are the results of testing.
Also @bismitaguha @sidvenu please review this again.

devkapilbansal commented:
@Aaishpra I haven't tested it yet. At first you have to squash your commits in one following the commit message guidelines. Then I will test it

Aaishpra reviewed (commented):
@rpattath since, i was having free time i tested my PR by myself (with the help of doc you provided on zulip ). I hope that's Ok. The PR already had two approvals by @bismitaguha and @sidvenu . But, since i had to squash my commits and remove spaces as suggested the approvals were canceled. So, @bismitaguha and @sidvenu please review this again.

The changes made in this PR were tested locally. Following are the results:

1. Code review - Done 

2. All possible responses (positive and negative tests) were tested as below:

   * _Test1 Description_  : Error Page should appear when the urls are wrong or broken
     _Screenshot/gif_: 
![image](https://user-images.githubusercontent.com/66299533/97069624-1d6d4700-15ef-11eb-881c-58d10cd254ec.png)
Gif link:https://drive.google.com/file/d/1V-h68CBC4jeAbMB4QFr_19J1IO-E0Izp/view?usp=sharing

     _Expected Result_:  The error page should appear.
     _Actual Result_: Same as Expected result.
   * _Test2 Description_  : UI should be responsive
     _Screenshot/gif_:  
![image](https://user-images.githubusercontent.com/66299533/97069672-602f1f00-15ef-11eb-977e-bd16c883dcf7.png)
Gif link: https://drive.google.com/file/d/1yHvXsRmFlJSikys8yfcZdoNJYsQkceaT/view?usp=sharing
     _Expected Result_:  UI is fully responsive.
     _Actual Result_:  same as expected result.
     ...  
  

4. Additional Comments: N/A


5. Status of PR Changed to: Needs Review.

khushishikhu commented:
I have checked this PR. This PR is ready to merge with the approval of maintainer. LGTM @Aaishpra

vj-codes reviewed (commented):
@Aaishpra this looks nice , here's a suggestion from my end -
It will be more user friendly if a button taking the user to the Home page is added below the error text , what say ? @isabelcosta 
If decided otherwise the PR is good to go:)

Aaishpra commented:
> @Aaishpra this looks nice , here's a suggestion from my end -
> It will be more user friendly if a button taking the user to the Home page is added below the error text , what say ? @isabelcosta 
> If decided otherwise the PR is good to go:)

There is no homepage yet in the application, wanted to add one but couldn't. 😆

vj-codes commented:
@Aaishpra ahh okay then 👀

vj-codes reviewed (dismissed):
LGTM 🎉

vj-codes commented:
@devkapilbansal hey can you review this ?

isabelcosta reviewed (dismissed):
@Aaishpra great work! Can't wait to see this on the website 🤗 
Thank you for this contribution!

vj-codes commented:
@isabelcosta this PR is tested above by @Aaishpra and all details are added as well
https://github.com/anitab-org/open-source-programs-web/pull/57#pullrequestreview-516164173

isabelcosta commented:
> this PR is tested above by @Aaishpra and all details are added as well

@vj-codes I see and I love it! However, when we think of testing PRs, it has to be someone who is not the author of the PR.

INNOVATIVEGAMER commented:
@Aaishpra You added custom css for all your styling. Styling is promising but this **project is already using Semantic-UI as UI framework**. I added some comments inside your css with some semantic-ui links which you can use in your component.
You can still use your custom one or two line css by adding `style={{}}` in components prop directly.
@isabelcosta I highly recommend you to add guidelines for usage of semantic-ui in designing of this repo.

isabelcosta commented:
@INNOVATIVEGAMER good point, could you please create a issue for us to add this to the guidelines please?
@Aaishpra could you take a look at this. How feasible it is to use this instead of custom styling? https://semantic-ui.com/

INNOVATIVEGAMER commented:
> @INNOVATIVEGAMER good point, could you please create a issue for us to add this to the guidelines please?

I created the issue for this #65. You can check on that. 
Also the contributing guidelines link is not working in readme with some mistakes, I created the issue for same also in #64 .

devkapilbansal reviewed (commented):
@vj-codes @isabelcosta I am not sure about coding practices and all.
But when I tested it locally, the UI changes looks good to me

The changes made in this PR is tested locally, following are the results:-
1) Code Review - *Not Done*

2) All possible responses were tested as below:
 
- Test1 Description - 404 error page appears when urls are wrong
Expected Result:-  The error page should appear
Actual Result:- Same as expected

- Test2 Description - The UI should be responsive
Actual Result same as expected
![Open Source Programs](https://user-images.githubusercontent.com/42766576/105640539-a7ee6c80-5ea4-11eb-98cd-cd298c7444ae.gif)

isabelcosta commented:
@devkapilbansal that is just awesome! thank you so much!

Aaishpra commented:
> @Aaishpra You added custom css for all your styling. Styling is promising but this **project is already using Semantic-UI as UI framework**. I added some comments inside your css with some semantic-ui links which you can use in your component.
> You can still use your custom one or two line css by adding `style={{}}` in components prop directly.
> @isabelcosta I highly recommend you to add guidelines for usage of semantic-ui in designing of this repo.

Thanks for the suggestion, I will update soon with the changes.

Aaishpra committed:
Fork update

Merge branch 'Aaishpra-patch-1' of https://github.com/Aaishpra/open-source-programs-web into Aaishpra-patch-1

Aaishpra committed:
semantic ui used

Aaishpra committed:
formatted document

Aaishpra committed:
Container removed

Aaishpra commented:
@isabelcosta I have modified the files according to the suggestion, without any change in designs. Here, I would like to mention one thing earlier I tried using `<Container Text>` from semantic-UI but it was breaking the design without centering the content, so I still have used some of my custom CSS to center the content and keep the designs same.

![image](https://user-images.githubusercontent.com/66299533/106983042-42935900-678b-11eb-8c0c-bcfcbb20f7d5.png)

Hope this is good.

codesankalp reviewed (commented):
Tested Locally @isabelcosta @Aaishpra @vj-codes 

Details of the test mentioned below:
**Code-Review: Not Done.**
**Tests:**
1. The 404 error page for invalid URL when not logged in.
     __Screenshot__:
     
![image](https://user-images.githubusercontent.com/56037184/107069064-685e4380-6807-11eb-8749-16d81ab3d831.png)

2. 404 Error page not showing when logged in it redirects to home page maybe this is due to AuthRoute.
__GIF__:
![captured gif](https://user-images.githubusercontent.com/56037184/107069761-6a74d200-6808-11eb-82cb-86694ce976fb.gif)

3. The scaling works perfect but for smaller width and height I observed this:
__Screenshot__:
![image](https://user-images.githubusercontent.com/56037184/107069968-adcf4080-6808-11eb-95db-fd831e7c05c3.png)

Also, Please squash your commits @Aaishpra .

isabelcosta commented:
I can squash so no need to squash @Aaishpra :)
@devkapilbansal and @codesankalp can you please add a code review?
@Aaishpra can you try to get more reviews from the community please, when it has 2 reviews I can merge :)

Rahulm2310 commented on src/styles/ErrorPage.css:
```diff
@@ -0,0 +1,26 @@
+}
+
+.errorPage {
+    position: absolute;
```
@Aaishpra Can I ask why are you using absolute positioning when this can be achieved with relative positioning as well. This way the overlapping issue on small screens mentioned by @codesankalp will also resolve.

Aaishpra replied:
` relative` is not gonna work on larger screen

Aaishpra commented:
This is what i see on my screen when changing ` position ` from `absolute` to ` relative ` (also the current design is working fine for smaller heights and width as tested by @devkapilbansal and i don't know how to make it work for even smaller widths)
![image](https://user-images.githubusercontent.com/66299533/107107511-7471f180-6857-11eb-9614-e295c6bf12b9.png)
about the` AuthRoute` part using of ` AuthRoute ` was suggested to me by @bismitaguha during code review of this PR (and i  also think that it's the correct way to do the redirect instead of using `Route`).

cc : @isabelcosta @codesankalp @Rahulm2310 @devkapilbansal

Rahulm2310 commented:
@Aaishpra Not just doing `position : relative` would work. Try adding `width :100% height:100vh`. As you already have `display flex, justify-content:center, align-items:center`, I think this should work. 🙂

Aaishpra commented:
> @Aaishpra Not just doing `position : relative` would work. Try adding `width :100% height:100vh`. As you already have `display flex, justify-content:center, align-items:center`, I think this should work. slightly_smiling_face

Not working

khushishikhu commented:
I have tested the PR. LGTM!! keep it up @Aaishpra

codesankalp commented on src/styles/ErrorPage.css:
```diff
@@ -0,0 +1,26 @@
+.errorPage_content p {
+    font-size: 1.4em;
+    color: gray;
+}
```
Can we use className (err-num and err-text) instead of `style={{}}` for Header in `ErrorPage.js` and add the styles in `ErrorPage.css` as mentioned below?
```suggestion
}

.ui.header.err-num {
    color: #2185D0;
    font-size: 10vw;
    font-weight: bolder;
    line-height: 1em;
    text-transform: uppercase;
}

.ui.header.err-text {
    font-size: 2em;
    margin-bottom: 20px;
    text-transform: uppercase;
    max-width: 600px;
    color: #0d0d0d;
}
```

isabelcosta replied:
Any specific reason @codesankalp ? Is this a common convention?

codesankalp reviewed (approved):
Code review is done. 
LGTM :+1:

isabelcosta commented:
Thank you all for the reviews :) I will merge it, and some comments about styling can be done at a later issue.
